File: Dished/View Controllers/DASettingsViewController.m
//
//  DASettingsViewController.m
//  Dished
//
//  Created by Ryan Khalili on 10/7/14.
//  Copyright (c) 2014 Dished. All rights reserved.
//

#import "DASettingsViewController.h"
#import "DAUserManager.h"
#import "DANotificationSettingsViewController.h"
#import "DAEditProfileViewController.h"
#import "DAAppDelegate.h"
#import "MRProgress.h"


@interface DASettingsViewController() <UIActionSheetDelegate>

@property (strong, nonatomic) NSURLSessionTask *profilePrivacyURLTask;

@end


@implementation DASettingsViewController

- (void)viewDidLoad
{
    [super viewDidLoad];
    
    self.tableView.rowHeight = 44.0;
    self.tableView.estimatedRowHeight = 44.0;
    
    [self populateSettings];
}

- (void)viewWillAppear:(BOOL)animated
{
    [super viewWillAppear:animated];
    
    [self.tableView deselectRowAtIndexPath:[self.tableView indexPathForSelectedRow] animated:YES];
}

- (void)populateSettings
{
    self.privacySwitch.on    = ![DAUserManager sharedManager].publicProfile;
    self.dishPhotosSwitch.on = [DAUserManager sharedManager].savesDishPhoto;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
    switch( indexPath.section )
    {
        case 0:
            [self handleAccountSectionSelectionForRow:indexPath.row];
            break;
            
        case 1:
            [self handleNotificationSectionSelectionForRow:indexPath.row];
            break;
            
        case 4:
            [self handleTermsSectionSelectionForRow:indexPath.row];
            break;
    }
}

- (void)handleAccountSectionSelectionForRow:(NSInteger)row
{
    switch( row )
    {
        case 0:
            [self goToEditProfile];
            break;
            
        case 1:
            [self performSegueWithIdentifier:@"shareSettings" sender:nil];
            break;
            
        case 2: 
            [self.tableView deselectRowAtIndexPath:[self.tableView indexPathForSelectedRow] animated:YES];
            [self showLogoutVerification];
            break;
    }
}

- (void)goToEditProfile
{
    DAEditProfileViewController *editProfileViewController = [self.storyboard instantiateViewControllerWithIdentifier:@"editProfile"];
    [self.navigationController pushViewController:editProfileViewController animated:YES];
}

- (void)showLogoutVerification
{
    UIActionSheet *actionSheet = [[UIActionSheet alloc] initWithTitle:@"Are you sure you want to Log Out?" delegate:self cancelButtonTitle:@"Cancel" destructiveButtonTitle:@"Log Out" otherButtonTitles:nil, nil];
    
    [actionSheet showInView:self.view];
}

- (void)actionSheet:(UIActionSheet *)actionSheet clickedButtonAtIndex:(NSInteger)buttonIndex
{
    if( buttonIndex == actionSheet.destructiveButtonIndex )
    {
        [MRProgressOverlayView showOverlayAddedTo:self.view.window title:@"Logging Out..." mode:MRProgressOverlayViewModeIndeterminate animated:YES];
        [self logout];
    }
}

- (void)logout
{
    [[DAAPIManager sharedManager] logoutWithCompletion:^( BOOL success )
    {
        [MRProgressOverlayView dismissOverlayForView:self.view.window animated:YES completion:^
        {
            if( success )
            {
                DAAppDelegate *appDelegate = (DAAppDelegate *)[[UIApplication sharedApplication] delegate];
                [appDelegate logout];
            }
            else
            {
                [[[UIAlertView alloc] initWithTitle:@"Failed to Log Out" message:@"There was a problem logging you out. Please try again." delegate:nil cancelButtonTitle:@"OK" otherButtonTitles:nil, nil] show];
            }
        }];
    }];
}

- (void)handleNotificationSectionSelectionForRow:(NSInteger)row
{
    NSString *notificationType = nil;
    
    switch( row )
    {
        case 0: notificationType = @"yum";     break;
        case 1: notificationType = @"comment"; break;
    }
    
    [self performSegueWithIdentifier:@"notifications" sender:notificationType];
}

- (void)handleTermsSectionSelectionForRow:(NSInteger)row
{
    switch( row )
    {
        case 0: [self goToDocumentViewWithName:kPrivacyPolicy]; break;
        case 1: [self goToDocumentViewWithName:kTermsAndConditions]; break;
    }
}

- (void)goToDocumentViewWithName:(NSString *)documentName
{
    NSString *filePath = [[NSBundle mainBundle] pathForResource:documentName ofType:@"html"];
    DADocViewController *docVC = [[DADocViewController alloc] initWithFilePath:filePath title:documentName];
    [self.navigationController pushViewController:docVC animated:YES];
}

- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender
{
    if( [segue.identifier isEqualToString:@"notifications"] )
    {
        DANotificationSettingsViewController *dest = segue.destinationViewController;
        dest.notificationType = sender;
    }
}

- (IBAction)changedProfilePrivacySetting
{
    [[DAUserManager sharedManager] savePrivacySetting:!self.privacySwitch.on completion:^( BOOL success )
    {
        [self populateSettings];
    }];
}

- (IBAction)changedDishPhotosSetting
{
    [[DAUserManager sharedManager] saveDishPhotoSetting:self.dishPhotosSwitch.on completion:^( BOOL success )
    {
        [self populateSettings];
    }];
}

@end

File: Dished/View Controllers/DASettingsViewController.swift
//
//  DASettingsViewController2.swift
//  Dished
//
//  Created by Ryan Khalili on 5/5/15.
//  Copyright (c) 2015 Dished. All rights reserved.
//

import UIKit

class DASettingsViewController: DAViewController, UITableViewDelegate, UITableViewDataSource, DASettingsTableViewCellDelegate, UIActionSheetDelegate {

    let settingsView = DASettingsView()
    let settingsDataSource = DASettingsDataSource()
    
    let cellIdentifier = "settingCell"
    
    override func viewDidLoad() {
        super.viewDidLoad()

        settingsView.tableView.delegate = self
        settingsView.tableView.dataSource = self
        settingsView.tableView.registerClass(DASettingsTableViewCell.self, forCellReuseIdentifier: cellIdentifier)
    }
    
    override func viewWillAppear(animated: Bool) {
        super.viewWillAppear(animated)
        
        settingsView.tableView.deselectSelectedIndexPath()
    }
    
    func numberOfSectionsInTableView(tableView: UITableView) -> Int {
        return settingsDataSource.sectionNames.count
    }
    
    func tableView(tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        let section = settingsDataSource.sectionNames[section]
        return settingsDataSource.settingsForSection(section).count
    }
    
    func tableView(tableView: UITableView, cellForRowAtIndexPath indexPath: NSIndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCellWithIdentifier(cellIdentifier) as! DASettingsTableViewCell
        
        let section = settingsDataSource.sectionNames[indexPath.section]
        let settings = settingsDataSource.settingsForSection(section)
        
        let style = settingsDataSource.cellStyleForSettingIndex(indexPath.row, inSection: section)
        cell.style = style
        cell.textLabel?.text = settings[indexPath.row]
        cell.delegate = self
        
        if style == DASettingsTableViewCellStyle.Switch {
            cell.selectorSwitch.on = settingsDataSource.stateForSetting(settings[indexPath.row])
        }
        
        return cell
    }
    
    func tableView(tableView: UITableView, didSelectRowAtIndexPath indexPath: NSIndexPath) {
        let section = settingsDataSource.sectionNames[indexPath.section]
        let settings = settingsDataSource.settingsForSection(section)
        let setting = settings[indexPath.row]
        
        if let destination = settingsDataSource.viewControllerForSetting(setting) {
            navigationController?.pushViewController(destination, animated: true)
        }
        else {
            tableView.deselectSelectedIndexPath()
            
            if settingsDataSource.settingIsLogout(setting) {
                showLogoutPrompt()
            }
        }
    }
    
    func tableView(tableView: UITableView, titleForHeaderInSection section: Int) -> String? {
        return settingsDataSource.sectionNames[section]
    }
    
    func tableView(tableView: UITableView, titleForFooterInSection section: Int) -> String? {
        return settingsDataSource.sectionFooters[section]
    }
    
    func settingsTableViewCell(cell: DASettingsTableViewCell, didSetSwitchOn on: Bool) {
        if let indexPath = settingsView.tableView.indexPathForView(cell.selectorSwitch) {
            let section = settingsDataSource.sectionNames[indexPath.section]
            let settings = settingsDataSource.settingsForSection(section)
            let setting = settings[indexPath.row]
            settingsDataSource.toggledSetting(setting, toState: on)
        }
    }
    
    private func showLogoutPrompt() {
        let actionSheet = UIActionSheet(title: "Are you sure you want to Log Out?", delegate: self, cancelButtonTitle: "Cancel", destructiveButtonTitle: "Log Out")
        actionSheet.showInView(view)
    }
    
    func actionSheet(actionSheet: UIActionSheet, didDismissWithButtonIndex buttonIndex: Int) {
        if buttonIndex == actionSheet.destructiveButtonIndex {
            logout()
        }
    }
    
    private func logout() {
        MRProgressOverlayView.showOverlayAddedTo(view, title: "Logging Out...", mode: MRProgressOverlayViewMode.Indeterminate, animated: true)
        
        settingsDataSource.logoutWithCompletion({
            success in
            
            if success {
                (UIApplication.sharedApplication().delegate as! DAAppDelegate).logout()
            }
            else {
                UIAlertView(title: "Failed to Log Out", message: "There was a problem logging you out. Please try again.", delegate: nil, cancelButtonTitle: "OK").show()
            }

        })
    }
    
    override func loadView() {
        view = settingsView
        
        navigationItem.title = "Settings"
    }
}